src/loop_in_chunks.cpp

```diff
@@ -185,44 +185,49 @@ typedef complex<double> cdouble;
 namespace meep {
 
 /***************************************************************/
-/* utility routine designed to be called by chunkloop functions*/
-/* for fetching values of field components at grid points,     */
-/* taking into account the complications of symmetry and       */
-/* yee-grid averaging.                                         */
-/* The first 5 arguments are the arguments of the same name    */
-/* passed to the chunkloop function.                           */
-/* The 6th argument is the integer idx defined inside the body */
-/* of a LOOP_OVER_IVECS.                                       */
-/* components is a caller-created list of components; the      */
-/* return value is an array of the same length storing the     */
-/* values of those components at the grid point.               */
+/* get_field_components is a utility routine, designed to be   */
+/* called by chunkloop functions, for fetching values of field */
+/* components at grid points, accounting for the complications */
+/* of symmetry and yee-grid averaging.                         */
 /***************************************************************/
-vector<complex<double> > get_field_components(fields_chunk *fc, component cgrid,
-                                              complex<double> shift_phase,
-                                              const symmetry &S, int sn,
-                                              int idx, vector<component> components)
-{
-  vector<complex<double> > fvals;
-
-  for (size_t nc=0; nc<components.size(); nc++)
-   { 
-     // get symmetry-parent component and yee-grid offsets
-     component cparent = S.transform(components[nc], -sn);
+field_component_data create_field_component_data(fields_chunk *fc, component cgrid,
+                                                 complex<double> shift_phase,
+                                                 const symmetry &S, int sn,
+                                                 vector <component> components)
```

Comment:
It feels odd to have a `create_foo(args...)` function in C++ rather than having a `foo` class with a `foo(args...)` constructor.